### PR TITLE
feat: Class for Fields with rank preserved by composition with conjTranspose Matrix

### PR DESCRIPTION
**Pull request by MohanadAhmed**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/Matrix/DotProductStarClass.lean [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+/-! # Types with Star Inner Product such that (star v) ⬝ᵥ v = 0 ↔ v = 0
+
+This file defines StarDotProductSpace in which the star operation on a vector makes the star
+dotProduct (in 1st arg) is zero if and only if the vector is zero. This makes star operation for
+matrices of the same underlying type kernel preserving for linear maps.
+-/
+
+open BigOperators Matrix
+
+class StarDotProductSpace (n K) [Fintype n][Ring K][StarRing K] : Prop where
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/DotProductStarClass.lean:23]: An alternative name for reviewers to consider
```suggestion
class DotProductInnerProductSpace (n K) [Fintype n][Ring K][StarRing K] : Prop where
```

File: Mathlib/LinearAlgebra/Matrix/DotProductStarClass.lean [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+variable {n : Type _}[Fintype n]
+
+lemma dotProduct_self_star_eq_zero  (v : n → K) [Ring K] [StarRing K]
+    [StarDotProductSpace n K] : Matrix.dotProduct v (star v) = 0 ↔ v = 0 := by
+  simpa only [star_star, star_eq_zero] using
+    StarDotProductSpace.dotProduct_star_self_eq_zero (star v)
+
+
+instance instStarDotProduct_IsROrC [IsROrC K] : StarDotProductSpace n K where
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/DotProductStarClass.lean:34]:
```suggestion
instance IsROrC.toStarDotProduct [IsROrC K] : StarDotProductSpace n K where
```

File: Mathlib/LinearAlgebra/Matrix/DotProductStarClass.lean [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+instance instQStar : Star ℚ where
+  star := fun x => x
+
+instance instTrivialQStar : TrivialStar ℚ where
+  star_trivial := by
+    intro x;
+    unfold star instQStar
+    simp only [Rat.cast_eq_id, id_eq]
+
+instance instQStarRing : StarRing ℚ where
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/DotProductStarClass.lean:71]: I'm pretty sure we already have this in another file.

File: Mathlib/LinearAlgebra/Matrix/DotProductStarClass.lean [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+This file defines StarDotProductSpace in which the star operation on a vector makes the star
+dotProduct (in 1st arg) is zero if and only if the vector is zero. This makes star operation for
+matrices of the same underlying type kernel preserving for linear maps.
+-/
+
+open BigOperators Matrix
+
+class StarDotProductSpace (n K) [Fintype n][Ring K][StarRing K] : Prop where
+  dotProduct_star_self_eq_zero (v : n → K) : Matrix.dotProduct (star v) v = 0 ↔ v = 0
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/DotProductStarClass.lean:24]:
```suggestion
  eq_zero_of_dotProduct_star_self_eq_zero (v : n → K) : Matrix.dotProduct (star v) v = 0 → v = 0
```
As the other direction is always true, so you don't need to prove it every time.

Comment (MohanadAhmed) [reply Mathlib/LinearAlgebra/Matrix/DotProductStarClass.lean]: I used this one at first but I am reverting it.

There is no saving coming from not proving it every time (in fact I needed to add `imp_self` every where.
In addition now the lemma cannot work as a rewrite.

Comment (eric-wieser) [reply Mathlib/LinearAlgebra/Matrix/DotProductStarClass.lean]: The point is that you write this simpler version in the `class`, and then prove the obvious `iff` as a lemma and use _that_ as the rewrite.

File: Mathlib/LinearAlgebra/Matrix/DotProductStarClass.lean [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      (fun _ => Complex.normSq_nonneg _)]
+
+
+instance Rat.toStarDotProduct : DotProductInnerProductSpace n ℚ where
+  dotProduct_star_self_eq_zero_iff := by
+    intro v
+    rw [star_trivial, dotProduct, Finset.sum_eq_zero_iff_of_nonneg, Function.funext_iff] <;>
+    simp only [Finset.mem_univ, mul_eq_zero, or_self, forall_true_left, Pi.zero_apply,
+      mul_self_nonneg, implies_true]
+
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/DotProductStarClass.lean:70]: Isn't this just a special case of the one below? Does
```
instance Rat.toStarDotProduct : DotProductInnerProductSpace n ℚ ;= inferInstance
```
work (at the bottom of the file)?

File: Mathlib/LinearAlgebra/Matrix/DotProductStarClass.lean [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+instance Complex.toStarDotProduct : DotProductInnerProductSpace n ℂ where
+  dotProduct_star_self_eq_zero_iff := by
+    intro v
+    simp_rw [dotProduct, Pi.star_apply, Complex.star_def, ← Complex.normSq_eq_conj_mul_self,
+      Complex.ext_iff, Complex.im_sum, Complex.zero_im, Complex.ofReal_im, Finset.sum_const_zero,
+      and_true, Complex.zero_re, Complex.re_sum, Complex.ofReal_re, Function.funext_iff,
+      Pi.zero_apply]
+    rw [Finset.sum_eq_zero_iff_of_nonneg] <;>
+    simp only [Finset.mem_univ, map_eq_zero, forall_true_left, implies_true,
+      (fun _ => Complex.normSq_nonneg _)]
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/DotProductStarClass.lean:61]: Does this work?
```suggestion
instance Real.toStarDotProduct : DotProductInnerProductSpace n ℝ := inferInstance

instance Complex.toStarDotProduct : DotProductInnerProductSpace n ℂ := inferInstance
```

Comment (MohanadAhmed) [reply Mathlib/LinearAlgebra/Matrix/DotProductStarClass.lean:61]: In the Real case you are correct. For the complex case, it does not work.
Maybe it needs an extra import

**Pull request closed without merging**